p4-16/spec/P4-16-spec.mdk

```diff
@@ -4412,6 +4414,33 @@ select (p.tcp.port) {
 The expression `16w0 &&& 16w0xFC00` describes the set of 16-bit
 values whose most significant six bits are zero.
 
+Some targets may support parser value set, see Section [#sec-value-set]. Given
+a type `T` for the type parameter of the value set, the type of the value set
+is `set<T>`. The type of the value set must match to the type of all other
+`keysetExpression` in the same `select` expression. If there is a mismatch, the
+compiler must raise an error. The type of the values in the set must be one of
+bit<>, tuple, and struct.
+
+For example, to allow control plane API to specify TCP reserved ports at
+runtime, one could write:
+
+~ Begin P4Example
+struct vsk_t {
+    @match(ternary)
+    bit<16> port;
```

Comment:
What if there is no @match annotation on a field?

Two possibilites I can think of:
(a) It is an error
(b) It defaults to @match(exact)

(b) seems reasonable to me, but whatever the choice, it would be good to say in the spec.

Is this syntax expected to be allowed?  `value_set<@match(ternary) bit<16>>(4) pvs;`  Or should we simply say that if you want a single non-exact field, you must put that field into a struct by itself, with the desired @match annotation?

Reply:
I modified the text to specify default is match(exact) and single non-exact field must be in a struct by itself.